stdlib/LinearAlgebra/src/generic.jl

```diff
@@ -728,16 +727,19 @@ julia> rank(Matrix(I, 3, 3))
 julia> rank(diagm(0 => [1, 0, 2]))
 2
 
-julia> rank(diagm(0 => [1, 0.001, 2]), 0.1)
+julia> rank(diagm(0 => [1, 0.001, 2]), rtol=0.1)
 2
 
-julia> rank(diagm(0 => [1, 0.001, 2]), 0.00001)
+julia> rank(diagm(0 => [1, 0.001, 2]), rtol=0.00001)
+3
+
+julia> rank(diagm(0 => [1, 0.001, 2]), atol=0.00001, rtol=0.00001)
 3
 ```
 """
-function rank(A::AbstractMatrix, tol::Real = min(size(A)...)*eps(real(float(one(eltype(A))))))
+function rank(A::AbstractMatrix; atol=0.0, rtol=0.0)
     s = svdvals(A)
-    count(x -> x > tol*s[1], s)
+    count(x -> x > (atol + rtol * s[1]), s)
```

Comment:
This should be made consistent with the current behavior of `isapprox`, which was changed in #22742 to use `max`: test `x > max(atol, rtol*s[1])`.

Reply:
(I don't know if the compiler can hoist a `max(atol, rtol*s[1])` call out of the `count` loop, but to be safe we should probably do it manually?  That is, set `tol = max(atol, rtol*s[1])` and test `x > tol`.

Reply:
A 0x0 matrix [is well defined](https://math.stackexchange.com/questions/1913141/is-the-0x0-matrix-zero-times-zero-matrix-a-well-defined-concept), and its rank should be zero (since its column space is 0-dimensional).  Similarly for any 0xN or Mx0 matrix.

So, I would add a line
```jl
isempty(A) && return 0 # 0-dimensional case
```
at the top of the function (before the call to `svdvals`).

Reply:
Or you could just leave it as `count(x -> x > max(atol, rtol*s[1]), s)`.  It's not like the performance matters for the `count` loop, since in any practical case it will be swamped by the cost of `svdvals`.